UBCFormulaElectric/Consolidated-Firmware
# Implement software watchdog to monitor each periodic task individually

### Summary

I implemented software watchdog to check for periodic task time. Once I implemented it I realized that the CAN TX library is performing so poorly to the point where the 1kHz task is missing its deadline (worst case execution time was 3ms). So I had to fix up the CAN TX library too.

The PR is longer than 200 words because I need to copy-paste the same code on each board.

### Changelist 

- Add drivers for software watchdog, and implement it on each board

### Testing Done

1. **Check watchdog with debugger connected:**

Uncomment each watchdog check-in, and observe that the automatic breakpoint is thrown and the `TASK_TIMEOUT` CAN message is sent.

2. **Check watchdog without  debugger connected:**

 `bkpt` has different behaviour, so we need to comment it out. In addition, if we want to see the periodic error CAN message update the `WATCHDOG_TIMEOUT` bit, then we have to change the prescaler to `IWDG`. This is because `IWDG` resets within 200ms but it takes 1000ms for the error message to update.

3. **Verify CAN TX task timing:**

With the new CAN design, the CAN TX task is lower priority than all periodic tasks. The advantage is that when there is a burst of CAN messages to be sent (e.g. during a RTOS tick that is a common multiple of CAN TX periods), we don't hog the CPU forever.

The image below illustrates how we send a burst of 8 messages in 1Khz task, after which the CAN TX tasks gradually sends out messages. Note that the 1kHz task is able to preempt CAN TX during the next RTOS tick (52002) which means the 1Khz task doesn't miss its deadline.
![image](https://user-images.githubusercontent.com/16970019/74294096-bd996f00-4cf1-11ea-88d3-822669fa480f.png)

I took care to make sure that we don't wake up the CAN TX task when there's nothing to be sent out. In the following picture, we see that there are 8 `xQueueReceive` corresponding to the 8 CAN messages that were enqueued earlier. I couldn't take a big enough screenshot to show how CAN TX task doesn't get woken up again after the 8th `xQeueuReceive`, but I've personally confirmed this. 

![image](https://user-images.githubusercontent.com/16970019/74294390-90998c00-4cf2-11ea-861b-4b43662f7eb3.png)

### Resolved Issues

Resolve #361 

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
